src/services/item/plugins/invitation/service.ts

```diff
@@ -0,0 +1,350 @@
+    await itemMembershipRepository.createMany(memberships);
+  }
+
+  async _partitionExistingUsersAndNewUsers(
+    actor: Actor,
+    repositories: Repositories,
+    emailList: string[],
+  ): Promise<{ existingAccounts: Member[]; newAccounts: string[] }> {
+    const { data: accounts } = await this.memberService.getManyByEmail(
+      actor,
+      repositories,
+      emailList,
+    );
+    const existingAccounts = Object.values(accounts);
+    const existingAccountsEmails = Object.keys(accounts);
+    const newAccounts = emailList.filter((email) => !existingAccountsEmails.includes(email));
+    return { existingAccounts, newAccounts };
+  }
+
+  async _createMembershipsAndInvitationsForUserList(
+    actor: Member,
+    repositories: Repositories,
+    rows: CSVInvite[],
+    itemId: Item['id'],
+    itemMembershipService: ItemMembershipService,
+  ) {
+    // partition between emails that are already accounts and emails without accounts
+    const { existingAccounts, newAccounts } = await this._partitionExistingUsersAndNewUsers(
+      actor,
+      repositories,
+      rows.map((r) => r.email),
+    );
+
+    // generate memberships to create
+    const membershipsToCreate = existingAccounts.map((account) => {
+      const permission =
+        // get the permission from the data, if it is not found or if it is an empty string, default to read
+        rows.find((r) => r.email === account.email)?.permission || PermissionLevel.Read;
+      return { permission, memberId: account.id };
+    });
+    this.log.debug(membershipsToCreate, 'memberships to create');
+
+    // create memberships for accounts that already exist
+    const memberships = await itemMembershipService.postMany(
+      actor,
+      repositories,
+      membershipsToCreate,
+      itemId,
+    );
+
+    // generate invitations to create
+    const invitationsToCreate = newAccounts.map((email) => {
+      // get the permission from the data, if it is not found or if it is an empty string, default to read
+      const permission = rows.find((r) => r.email === email)?.permission || PermissionLevel.Read;
+      return { email, permission };
+    });
+    this.log.debug(invitationsToCreate, 'invitations to create');
+
+    // create invitations for accounts that do not exist yet
+    const invitations = await this.postManyForItem(
+      actor,
+      repositories,
+      itemId,
+      invitationsToCreate,
+    );
+
+    return { memberships, invitations };
+  }
+
+  async importUsersWithCSV(
+    actor: Member,
+    repositories: Repositories,
+    itemId: Item['id'],
+    file: MultipartFile,
+    itemMembershipService: ItemMembershipService,
+  ): Promise<{ memberships: ItemMembership[]; invitations: Invitation[] }> {
+    // verify file is CSV
+    verifyCSVFileFormat(file);
+
+    // get parentItem
+    const item = await repositories.itemRepository.get(itemId);
+    await validatePermission(repositories, PermissionLevel.Admin, actor, item);
```

Comment:
If you get the item service, you can use `itemService.get()` that validates the permission for you. Also you can request for a specific permission level.

Reply:
Okay will use that !